equiv check (#99)
Certora/Documentation: docs/equiv-check/index.md

(equivcheck-intro)=
Equivalence Checking Using the Certora Prover
=================================

> This chapter describes how one can use the Certora Prover to
>   check the equivalence of two pure functions.
> 
> ```{note}
> We are working on extending this to support functions with side effects;
>   for the remaining documentation,
>   we assume a user is interested in running the equivalence checker
>   for two pure functions which may belong to the same contract
>   or two different contracts.
> ```

Reviewer: ````suggestion
This chapter describes how one can use the Certora Prover to
  check the equivalence of two smart contract functions.

```{note}
Currently the equivalence checker only compares two `pure` functions, but we are actively working to develop an equivalence checker for non-`pure` functions as well.
```
````

The equivalence checker front-end automatically generates (1) a
  CVL spec to check if two functions are equivalent, and, (2) a

>   configuration file (`.conf`) for running the Certora Prover.

Reviewer: this should be a link but there's nothing to link to :(

## Installation

> The front-end script, `certoraEqCheck`,
>   should be available as part installing `certora-cli`.

Reviewer: ```suggestion
The equivalence checker is part of the `certora-cli` package; see {ref}`installation`.
```

(example)=
## Example

> Consider two contracts, `BasicMathGood.sol` and `BasicMathBad.sol` shown
>   below with two functions, `add` and `add_mult`.

Reviewer: ```suggestion
Consider two contracts, `BasicMathGood.sol` and `BasicMathBad.sol` shown
  below with two functions, `add` and `add_mult`.
```

Author: i dont see a diff here

> ```
> BasicMathGood.sol:
> ------------------
> 
> contract BasicMathGood {
>     function add(uint256 a, uint256 b) public pure returns(uint256) {
>         return a + b;
>     }
> }
> 
> ```
> 
> ```
> BasicMathBad.sol:
> ------------------
> 
> contract BasicMathBad {
>     function add_mult(uint256 a, uint256 b) public pure returns(uint256) {
>         return a * b;
>     }
> }
> ```

Reviewer: ````suggestion
```solidity
contract BasicMathGood {
    function add(uint256 a, uint256 b) public pure returns(uint256) {
        return a + b;
    }
}

contract BasicMathBad {
    function add_mult(uint256 a, uint256 b) public pure returns(uint256) {
        return a * b;
    }
}
```
````

> We are interested in checking the equivalence of `add` and `add_mult`.

Reviewer: might be nice to do an example with an `unchecked` block, or maybe even a line or two of assembly to sell it.  Then the functions won't "clearly" be non-equivalent.

 Certora Prover `conf` file along with additional arguments.

### Default mode

To run the equivalence checker in default mode,
  use `certoraEqCheck` like so:

```bash
certoraEqCheck def "path_to_file:contract:function:solc" "path_to_file:contract:function:solc"
```

For the functions in {ref}`example`, this would look like so:

```bash
certoraEqCheck def Test/EqCheck/BasicMathGood.sol:add:solc8.0 Test/EqCheck/BasicMathBad.sol:add_pass:solc8.0
```

In the above example, `solc` is the name of the executable
  for the Solidity compiler version you are using.
The Solidity compilers do not need to be the same for both arguments to
 `certoraEqCheck`, it only need to be appropriate for the given contract.
Also note how
  the contract field can be omitted if the contract name is the same
  the file name.

### Configuration mode

To run the equivalence checker in the configuration mode,
  use `certoraEqCheck` like so:

```bash
certoraEqCheck conf <path_to_conf>.conf contract:function contract:function
```

For the functions in {ref}`example`, this would look like so:

```bash
   certoraEqCheck conf Test/EqCheck/testGood.conf BasicMathGood:add BasicMathBad:add_mult
```

> where `testGood.conf` contains the following configuration:

Reviewer: Is this a standard `certoraRun` config file?  If so, it's probably worth saying that.  I would normally link to the documentation, except we don't have it right now. It might be worth creating a placeholder page with a {todo} and linking to that.  Maybe ask @rahav-certora - I think he maybe is working on one?

Author: ok i asked him and added a clarification for now

(equivcheck-intro)=

> 

Reviewer: Missing title is currently breaking the build
